File: cloud-run/modules/ai_router.py
"""
AIルーターモジュール
Gemini 3 Flash優先、信頼度スコアに基づくProへのエスカレーション
"""
import json
import logging
from typing import Dict, Optional
from google.cloud import secretmanager
import google.generativeai as genai
from config.settings import (
    GEMINI_MODELS, 
    AI_ROUTER_CONFIG,
    GCP_PROJECT_ID,
    SECRET_GEMINI_API_KEY
)

logger = logging.getLogger(__name__)


class AIRouter:
    """Gemini 3 Flash/Proを使い分けるAIルーター"""
    
    def __init__(self):
        """初期化"""
        self.api_key = self._get_api_key()
        genai.configure(api_key=self.api_key)
        
    def _get_api_key(self) -> str:
        """Secret ManagerからGemini APIキーを取得"""
        try:
            client = secretmanager.SecretManagerServiceClient()
            name = f"projects/{GCP_PROJECT_ID}/secrets/{SECRET_GEMINI_API_KEY}/versions/latest"
            response = client.access_secret_version(request={"name": name})
            return response.payload.data.decode('UTF-8').strip()
        except Exception as e:
            logger.error(f"Secret Manager APIキー取得エラー: {e}")
            # フォールバック: 環境変数から取得
            import os
            api_key = os.getenv('GEMINI_API_KEY')
            if not api_key:
                raise ValueError("Gemini API keyが見つかりません")
            return api_key.strip()
    
    def analyze_document(
        self, 
        image_data: bytes, 
        prompt: str,
        use_flash_first: bool = True
    ) -> Optional[Dict]:
        """
        ドキュメントを解析（AIルーターパターン）
        
        Args:
            image_data: 画像バイナリデータ
            prompt: プロンプト
            use_flash_first: Flash優先フラグ
            
        Returns:
            解析結果（JSON）
        """
        if use_flash_first:
            # 第一段階: Gemini 3 Flash
            logger.info("Gemini 3 Flashで解析開始")
            flash_result = self._call_gemini(
                GEMINI_MODELS['FLASH'],
                image_data,
                prompt
            )
            
            if flash_result and self._is_confident(flash_result):
                logger.info(f"Flash解析成功（信頼度: {flash_result.get('confidence_score', 'N/A')}）")
                return flash_result
            
            # 第二段階: Gemini 3 Proへエスカレーション
            if AI_ROUTER_CONFIG['ENABLE_PRO_ESCALATION']:
                logger.warning(
                    f"信頼度が低いためProにエスカレーション "
                    f"(score: {flash_result.get('confidence_score', 0.0) if flash_result else 'N/A'})"
                )
                return self._call_gemini(
                    GEMINI_MODELS['PRO'],
                    image_data,
                    prompt
                )
            
            return flash_result
        else:
            # Pro直接呼び出し
            return self._call_gemini(
                GEMINI_MODELS['PRO'],
                image_data,
                prompt
            )
    
    def _call_gemini(
        self, 
        model_name: str, 
        image_data: bytes, 
        prompt: str
    ) -> Optional[Dict]:
        """Gemini APIを呼び出し"""
        try:
            # デバッグ: APIキーの確認（先頭/末尾のみ表示）
            if self.api_key:
                masked_key = f"{self.api_key[:4]}...{self.api_key[-4:]}"
                logger.info(f"Using API Key: {masked_key} (Length: {len(self.api_key)})")
            else:
                logger.error("API Key is empty/None")

            model = genai.GenerativeModel(
                model_name,
                generation_config={
                    "response_mime_type": "application/json"
                }
            )
            
            # 画像とプロンプトを送信
            import base64
            image_b64 = base64.b64encode(image_data).decode('utf-8')
            
            response = model.generate_content([
                {
                    "mime_type": "image/jpeg",
                    "data": image_b64
                },
                prompt
            ])
            
            # JSON形式でパース
            result = json.loads(response.text)
            return result
            
        except Exception as e:
            # repr()を使ってエンコーディング問題を回避
            logger.error(f"Gemini API呼び出しエラー ({model_name}): {repr(e)}")
            import traceback
            logger.error(f"Traceback: {traceback.format_exc()}")
            return None
    
    def _is_confident(self, result: Dict) -> bool:
        """信頼度スコアをチェック"""
        if not result:
            return False
        confidence = result.get('confidence_score', 0.0)
        threshold = AI_ROUTER_CONFIG['CONFIDENCE_THRESHOLD']
        return confidence >= threshold

    def extract_events_and_tasks(
        self,
        image_data: bytes,
        file_name: str
    ) -> Optional[Dict]:
        """
        ドキュメントから予定とタスクを抽出
        
        Args:
            image_data: 画像バイナリデータ
            file_name: ファイル名
            
        Returns:
            抽出結果（JSON）
        """
        from datetime import datetime
        today = datetime.now().strftime('%Y-%m-%d')
        
        prompt = f"""
あなたは学校のお便りから予定とタスクを抽出するアシスタントです。
以下の画像を解析し、JSON形式で回答してください。

## 出力形式（必ずこのJSON形式で回答）
{{
  "events": [
    {{
      "title": "イベントタイトル",
      "date": "YYYY-MM-DD",
      "start_time": "HH:MM（不明な場合は null）",
      "end_time": "HH:MM（不明な場合は null）",
      "location": "場所（不明な場合は null）",
      "description": "詳細説明"
    }}
  ],
  "tasks": [
    {{
      "title": "タスクタイトル（例：○○の提出）",
      "due_date": "YYYY-MM-DD",
      "notes": "備考"
    }}
  ]
}}

## 判断基準
- **events**: 日時が確定している行事（運動会、授業参観、保護者会など）
- **tasks**: 期限がある提出物や準備事項（書類提出、持ち物準備など）

## 注意事項
- 過去の日付（{today}より前）のイベント・タスクは除外してください
- 年が明示されていない場合は、{today[:4]}年と仮定してください
- 抽出できる情報がない場合は、eventsとtasksを空配列にしてください

## ファイル名
{file_name}
"""
        # 構造化抽出のためProモデル使用またはFlashでもJSONモード
        # ここでは既存の _call_gemini を使用 (Flash優先)
        return self.analyze_document(image_data, prompt, use_flash_first=True)

File: cloud-run-go/internal/service/ai_router.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	secretmanager "cloud.google.com/go/secretmanager/apiv1"
	"cloud.google.com/go/secretmanager/apiv1/secretmanagerpb"
	"github.com/google/generative-ai-go/genai"
	"github.com/leo-sagawa/homedocmanager/internal/config"
	"github.com/leo-sagawa/homedocmanager/internal/model"
	"google.golang.org/api/option"
)

// AIRouter はGemini Flash/Proを使い分けるAIルーター
type AIRouter struct {
	apiKey string
	client *genai.Client
}

// NewAIRouter は新しいAIRouterを作成
func NewAIRouter(ctx context.Context) (*AIRouter, error) {
	apiKey, err := getAPIKey(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to get API key: %w", err)
	}

	client, err := genai.NewClient(ctx, option.WithAPIKey(apiKey))
	if err != nil {
		return nil, fmt.Errorf("failed to create genai client: %w", err)
	}

	return &AIRouter{
		apiKey: apiKey,
		client: client,
	}, nil
}

// getAPIKey はSecret ManagerからGemini APIキーを取得
func getAPIKey(ctx context.Context) (string, error) {
	// Secret Managerクライアントを作成
	client, err := secretmanager.NewClient(ctx)
	if err != nil {
		// フォールバック: 環境変数から取得
		log.Printf("Secret Manager client creation failed: %v, falling back to env var", err)
		return getFallbackAPIKey()
	}
	defer client.Close()

	// Secret Managerからキーを取得
	name := fmt.Sprintf("projects/%s/secrets/%s/versions/latest",
		config.GCPProjectID,
		config.SecretGeminiAPIKey)

	req := &secretmanagerpb.AccessSecretVersionRequest{
		Name: name,
	}

	result, err := client.AccessSecretVersion(ctx, req)
	if err != nil {
		// フォールバック: 環境変数から取得
		log.Printf("Secret Manager access failed: %v, falling back to env var", err)
		return getFallbackAPIKey()
	}

	return string(result.Payload.Data), nil
}

// getFallbackAPIKey は環境変数からAPIキーを取得
func getFallbackAPIKey() (string, error) {
	apiKey := config.GetEnv("GEMINI_API_KEY", "")
	if apiKey == "" {
		return "", fmt.Errorf("GEMINI_API_KEY not found in environment variables")
	}
	return apiKey, nil
}

// AnalyzeDocument はドキュメントを解析（AIルーターパターン）
func (r *AIRouter) AnalyzeDocument(ctx context.Context, data []byte, mimeType string, prompt string, useFlashFirst bool) (*model.AnalysisResult, error) {
	if useFlashFirst {
		// 第一段階: Gemini Flash
		log.Println("Gemini Flash で解析開始")
		flashResult, err := r.callGemini(ctx, config.GeminiModelsConfig.Flash, data, mimeType, prompt)
		if err == nil && r.isConfident(flashResult) {
			log.Printf("Flash解析成功（信頼度: %.2f）", flashResult.ConfidenceScore)
			return flashResult, nil
		}

		// 第二段階: Gemini Proへエスカレーション
		if config.AIRouter.EnableProEscalation {
			score := 0.0
			if flashResult != nil {
				score = flashResult.ConfidenceScore
			}
			log.Printf("信頼度が低いためProにエスカレーション (score: %.2f)", score)
			return r.callGemini(ctx, config.GeminiModelsConfig.Pro, data, mimeType, prompt)
		}

		return flashResult, err
	}

	// Pro直接呼び出し
	return r.callGemini(ctx, config.GeminiModelsConfig.Pro, data, mimeType, prompt)
}

// callGemini はGemini APIを呼び出し
func (r *AIRouter) callGemini(ctx context.Context, modelName string, data []byte, mimeType string, prompt string) (*model.AnalysisResult, error) {
	genModel := r.client.GenerativeModel(modelName) // Rename variable to avoid shadowing package name
	genModel.GenerationConfig.ResponseMIMEType = "application/json"

	// 画像/PDFとプロンプトを送信
	// mimeTypeがimage/jpeg等の場合はImageDataでも良いが、汎用的にBlobを使用可能か確認
	// genai.ImageDataは shortcut for Blob(mimeType, data) with validation for image types?
	// PDFの場合はBlobを使う必要がある。

	var dataPart genai.Part
	if mimeType == "application/pdf" {
		dataPart = genai.Blob{
			MIMEType: mimeType,
			Data:     data,
		}
	} else {
		// 画像（JPEG, PNG等）
		// mimeTypeから拡張子部分を取得（簡易的）
		format := "jpeg"
		if len(mimeType) > 6 {
			format = mimeType[6:]
		}
		dataPart = genai.ImageData(format, data)
	}

	resp, err := genModel.GenerateContent(ctx,
		dataPart,
		genai.Text(prompt),
	)
	if err != nil {
		return nil, fmt.Errorf("gemini API call failed (%s): %w", modelName, err)
	}

	// JSONレスポンスをパース
	if len(resp.Candidates) == 0 || len(resp.Candidates[0].Content.Parts) == 0 {
		return nil, fmt.Errorf("no response from gemini API")
	}

	// レスポンステキストを取得
	text := fmt.Sprintf("%v", resp.Candidates[0].Content.Parts[0])

	var result model.AnalysisResult
	if err := json.Unmarshal([]byte(text), &result); err != nil {
		return nil, fmt.Errorf("failed to parse JSON response: %w", err)
	}

	return &result, nil
}

// isConfident は信頼度スコアをチェック
func (r *AIRouter) isConfident(result *model.AnalysisResult) bool {
	if result == nil {
		return false
	}
	return result.ConfidenceScore >= config.AIRouter.ConfidenceThreshold
}

// ExtractEventsAndTasks はドキュメントから予定とタスクを抽出
func (r *AIRouter) ExtractEventsAndTasks(ctx context.Context, data []byte, mimeType string, fileName string) (*model.EventsAndTasks, error) {
	today := time.Now().Format("2006-01-02") // Use time.Now()

	prompt := fmt.Sprintf(`
あなたは学校のお便りから予定とタスクを抽出するアシスタントです。
以下の画像を解析し、JSON形式で回答してください。

## 出力形式（必ずこのJSON形式で回答）
{
  "events": [
    {
      "title": "イベントタイトル",
      "date": "YYYY-MM-DD",
      "start_time": "HH:MM（不明な場合は null）",
      "end_time": "HH:MM（不明な場合は null）",
      "location": "場所（不明な場合は null）",
      "description": "詳細説明"
    }
  ],
  "tasks": [
    {
      "title": "タスクタイトル（例：○○の提出）",
      "due_date": "YYYY-MM-DD",
      "notes": "備考"
    }
  ]
}

## 判断基準
- **events**: 日時が確定している行事（運動会、授業参観、保護者会など）
- **tasks**: 期限がある提出物や準備事項（書類提出、持ち物準備など）

## 注意事項
- 過去の日付（%sより前）のイベント・タスクは除外してください
- 年が明示されていない場合は、%s年と仮定してください
- 抽出できる情報がない場合は、eventsとtasksを空配列にしてください

## ファイル名
%s
`, today, today[:4], fileName)

	genModel := r.client.GenerativeModel(config.GeminiModelsConfig.Flash) // Rename to genModel
	genModel.GenerationConfig.ResponseMIMEType = "application/json"

	var dataPart genai.Part
	if mimeType == "application/pdf" {
		dataPart = genai.Blob{
			MIMEType: mimeType,
			Data:     data,
		}
	} else {
		format := "jpeg"
		if len(mimeType) > 6 {
			format = mimeType[6:]
		}
		dataPart = genai.ImageData(format, data)
	}

	resp, err := genModel.GenerateContent(ctx,
		dataPart,
		genai.Text(prompt),
	)
	if err != nil {
		return nil, fmt.Errorf("gemini API call failed: %w", err)
	}

	if len(resp.Candidates) == 0 || len(resp.Candidates[0].Content.Parts) == 0 {
		return nil, fmt.Errorf("no response from gemini API")
	}

	text := fmt.Sprintf("%v", resp.Candidates[0].Content.Parts[0])

	var result model.EventsAndTasks
	if err := json.Unmarshal([]byte(text), &result); err != nil {
		return nil, fmt.Errorf("failed to parse JSON response: %w", err)
	}

	return &result, nil
}

// ExtractOCRText はドキュメントからプレーンテキストを抽出（OCR）- 互換性維持用ラッパー
func (r *AIRouter) ExtractOCRText(ctx context.Context, data []byte, mimeType string) (string, error) {
	bundle, err := r.ExtractOCRBundle(ctx, data, mimeType)
	if err != nil {
		return "", err
	}
	return bundle.OCRText, nil
}

// ExtractOCRBundle はドキュメントからOCRテキスト、事実、要約を構造化して抽出
func (r *AIRouter) ExtractOCRBundle(ctx context.Context, data []byte, mimeType string) (*model.OCRBundle, error) {
	prompt := `この画像/ドキュメントを解析し、以下のJSON形式で回答してください。

{
  "ocr_text": "画像/ドキュメントに含まれるすべてのテキスト（OCR結果）",
  "facts": [
    "実際に本文から抽出した項目のみ"
  ],
  "summary": "要約（1-2文、曖昧な場合は空文字）",
  "confidence_score": 0.0-1.0,
  "quality": {
    "uncertain": false,
    "needs_high_model": false,
    "notes": "読み取りに問題があれば記載"
  }
}

重要ルール：
- factsには実際に本文から抽出した項目のみを入れてください。説明文やテンプレ文は絶対に含めないでください。
- factsには日付、金額、名前、住所など具体的な情報を抽出。推測は禁止。
- 見つからない場合は "不明" を入れてください。最大10項目。
- 読み取れない部分がある場合は quality.uncertain を true に。
- summaryは確信がなければ空文字で可。`

	genModel := r.client.GenerativeModel(config.GeminiModelsConfig.Flash)
	genModel.GenerationConfig.ResponseMIMEType = "application/json"

	var dataPart genai.Part
	if mimeType == "application/pdf" {
		dataPart = genai.Blob{
			MIMEType: mimeType,
			Data:     data,
		}
	} else {
		format := "jpeg"
		if len(mimeType) > 6 {
			format = mimeType[6:]
		}
		dataPart = genai.ImageData(format, data)
	}

	resp, err := genModel.GenerateContent(ctx,
		dataPart,
		genai.Text(prompt),
	)
	if err != nil {
		return nil, fmt.Errorf("OCR extraction failed: %w", err)
	}

	if len(resp.Candidates) == 0 || len(resp.Candidates[0].Content.Parts) == 0 {
		return &model.OCRBundle{}, nil
	}

	// Partsを連結してテキストを取得
	text := r.extractTextFromParts(resp.Candidates[0].Content.Parts)

	var bundle model.OCRBundle
	if err := json.Unmarshal([]byte(text), &bundle); err != nil {
		// JSONパース失敗時はプレーンテキストとして扱う
		log.Printf("OCRBundle JSONパース失敗、プレーンテキストとして処理: %v", err)
		bundle.OCRText = text
		bundle.ConfidenceScore = 0.5
	}

	// Flash結果が弱い場合はProにエスカレーション
	if bundle.ConfidenceScore < config.AIRouter.ConfidenceThreshold ||
		bundle.Quality.NeedsHighModel ||
		(bundle.Quality.Uncertain && len(bundle.Facts) < 3) ||
		len(bundle.OCRText) < 200 {
		if config.AIRouter.EnableProEscalation {
			log.Printf("OCR精度が低いためProにエスカレーション (score: %.2f, len: %d)", bundle.ConfidenceScore, len(bundle.OCRText))
			return r.extractOCRBundleWithModel(ctx, data, mimeType, config.GeminiModelsConfig.Pro, prompt)
		}
	}

	log.Printf("OCRBundle抽出完了: %d文字, %d facts", len(bundle.OCRText), len(bundle.Facts))
	return &bundle, nil
}

// extractTextFromParts はPartsからテキストを安全に抽出
func (r *AIRouter) extractTextFromParts(parts []genai.Part) string {
	var result string
	for _, part := range parts {
		switch v := part.(type) {
		case genai.Text:
			result += string(v)
		default:
			// Text以外のPartは文字列化
			result += fmt.Sprintf("%v", v)
		}
	}
	return result
}

// extractOCRBundleWithModel は指定モデルでOCRBundle抽出
func (r *AIRouter) extractOCRBundleWithModel(ctx context.Context, data []byte, mimeType, modelName, prompt string) (*model.OCRBundle, error) {
	genModel := r.client.GenerativeModel(modelName)
	genModel.GenerationConfig.ResponseMIMEType = "application/json"

	var dataPart genai.Part
	if mimeType == "application/pdf" {
		dataPart = genai.Blob{
			MIMEType: mimeType,
			Data:     data,
		}
	} else {
		format := "jpeg"
		if len(mimeType) > 6 {
			format = mimeType[6:]
		}
		dataPart = genai.ImageData(format, data)
	}

	resp, err := genModel.GenerateContent(ctx,
		dataPart,
		genai.Text(prompt),
	)
	if err != nil {
		return nil, fmt.Errorf("OCR extraction (Pro) failed: %w", err)
	}

	if len(resp.Candidates) == 0 || len(resp.Candidates[0].Content.Parts) == 0 {
		return &model.OCRBundle{}, nil
	}

	// Partsを連結してテキストを取得
	text := r.extractTextFromParts(resp.Candidates[0].Content.Parts)

	var bundle model.OCRBundle
	if err := json.Unmarshal([]byte(text), &bundle); err != nil {
		bundle.OCRText = text
		bundle.ConfidenceScore = 0.5
	}

	log.Printf("OCRBundle抽出完了 (Pro): %d文字, %d facts", len(bundle.OCRText), len(bundle.Facts))
	return &bundle, nil
}

// Close はクライアントをクローズ
func (r *AIRouter) Close() error {
	return r.client.Close()
}
